File: src/main/java/kr/syeyoung/dungeonsguide/mod/whosonline/api/WhosOnlineApi.java
package kr.syeyoung.dungeonsguide.mod.whosonline.api;

import kr.syeyoung.dungeonsguide.mod.stomp.StompClient;
import kr.syeyoung.dungeonsguide.mod.whosonline.WhosOnlineManager;
import kr.syeyoung.dungeonsguide.mod.whosonline.api.messages.client.C02IsOnline;
import kr.syeyoung.dungeonsguide.mod.whosonline.api.messages.client.C04OnlineCheckBulk;
import lombok.val;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;

/**
 * Api wrapper for whos online websocket server
 * @author Eryk Ruta
 */
public class WhosOnlineApi {
    private static final long TIMEOUT_VALUE = 500;
    private final WhosOnlineWebSocket client;
    private final ExecutorService executor;

    public WhosOnlineApi(WhosOnlineWebSocket client, ExecutorService ex) {
        this.client = client;

        executor = ex;
    }

    Logger logger = LogManager.getLogger("WhosOnlineApi");


    /**
     * @return -s if websocket is ok
     */
    boolean stateCheck() {
        return client.getStatus() != StompClient.StompClientStatus.CONNECTED;
    }

    /**
     * @param uuid UUid of player to check
     * @return true if player is online with dg
     */
    public @Nullable Future<Boolean> isOnline(@NotNull final String uuid) {
        if (stateCheck()) return null;
        return executor.submit(() -> {
            val messageId = UUID.randomUUID().toString();
            val message = WhosOnlineManager.gson.toJson(new C02IsOnline(new C02IsOnline.OBJ(uuid, messageId)));

            client.sendAndBlock(message, messageId, TIMEOUT_VALUE);

            logger.info("Is online: {} uuid: {} ", uuid , WhosOnlineCache.onlineppl.getOrDefault(uuid, false));

            return WhosOnlineCache.onlineppl.getOrDefault(uuid, false);

        });
    }

    /**
     * @param uuids uuids of player to check
     * @return array of their statuses
     */
    public @Nullable Future<Boolean[]> areOnline(@NotNull final String[] uuids) {
        if (stateCheck()) return null;
        return executor.submit(() -> {
            val messageId = UUID.randomUUID().toString();
            val message = WhosOnlineManager.gson.toJson(new C04OnlineCheckBulk(new C04OnlineCheckBulk.OBJ(uuids, messageId)));

            client.sendAndBlock(message, messageId, TIMEOUT_VALUE);

            val returnVals = new Boolean[uuids.length];

            for (int i = 0; i < uuids.length; i++) {
                val uuid = uuids[i];
                returnVals[i] = WhosOnlineCache.onlineppl.getOrDefault(uuid, false);
            }

            return returnVals;

        });
    }


}

File: src/main/java/kr/syeyoung/dungeonsguide/whosonline/api/WhosOnlineApi.kt
package kr.syeyoung.dungeonsguide.whosonline.api

import kr.syeyoung.dungeonsguide.whosonline.WebSocketClientStatus
import kr.syeyoung.dungeonsguide.whosonline.WhosOnlineManager
import kr.syeyoung.dungeonsguide.whosonline.api.messages.client.C01IsOnline
import kr.syeyoung.dungeonsguide.whosonline.api.messages.client.C02OnlineCheckBulk
import org.apache.logging.log4j.LogManager
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

/**
 * Api wrapper for whos online websocket server
 * @author Eryk Ruta
 */
class WhosOnlineApi(
    private val client: WhosOnlineWebSocket,
    private val cache: WhosOnlineCache,
    private val executor: ExecutorService
) {
    var logger = LogManager.getLogger("WhosOnlineApi")

    /**
     * @return -s if websocket is ok
     */
    fun stateCheck(): Boolean {
        return client.status != WebSocketClientStatus.CONNECTED
    }

    /**
     * @param uuid UUid of player to check
     * @return true if player is online with dg
     */
    fun isOnline(uuid: String): Future<Boolean> {
        if (cache.isCached(uuid)) {
            return CompletableFuture.completedFuture(cache.isOnline(uuid))
        }
        return if (stateCheck()) {
            CompletableFuture.completedFuture(false)
        } else executor.submit<Boolean> {
            val messageId = UUID.randomUUID().toString()
            val message = WhosOnlineManager.gson.toJson(C01IsOnline(C01IsOnline.OBJ(uuid, messageId)))
            client.sendAndBlock(message, messageId, TIMEOUT_VALUE)
            logger.info("Is online: {} uuid: {} ", uuid, cache.isOnline(uuid))
            cache.isOnline(uuid)
        }
    }

    /**
     * @param uuids uuids of player to check
     * @return array of their statuses
     */
    fun areOnline(uuids: Array<String>): Future<Array<Boolean?>> {
        if (stateCheck()) {
            val nulls = arrayOfNulls<Boolean>(uuids.size)
            Arrays.fill(nulls, false)
            return CompletableFuture.completedFuture(nulls)
        }
        val notCached: MutableSet<String> = HashSet()
        val cached = HashMap<String, Boolean>()
        for (uuid in uuids) {
            val iscached = cache.isCached(uuid)
            if (iscached) {
                notCached.add(uuid)
            } else {
                cached[uuid] = cache.isOnline(uuid)
            }
        }

        // in case that all the nicks are not cached
        if (notCached.isEmpty()) {
            val res = arrayOfNulls<Boolean>(uuids.size)
            // we do this to preserve the order which they were added in
            for (i in uuids.indices) {
                val uuid = uuids[i]
                res[i] = cached[uuid]
            }
            return CompletableFuture.completedFuture(res)
        }
        return executor.submit<Array<Boolean?>> {
            val messageId = UUID.randomUUID().toString()
            val message = WhosOnlineManager.gson.toJson(
                C02OnlineCheckBulk(
                    C02OnlineCheckBulk.OBJ(
                        messageId,
                        notCached.toTypedArray(),
                    )
                )
            )
            client.sendAndBlock(message, messageId, TIMEOUT_VALUE)
            val returnVals = arrayOfNulls<Boolean>(uuids.size)
            for (i in uuids.indices) {
                val uuid = uuids[i]
                returnVals[i] = cache.isOnline(uuid)
            }
            returnVals
        }
    }

    companion object {
        private const val TIMEOUT_VALUE: Long = 500
    }
}